pypreprocess/realign.py

```diff
@@ -300,112 +461,33 @@ def _single_session_fit(self, vols, quality=None,
-                y1, y2, y3 = transform_coords(np.zeros(6), vol_0.get_affine(),
-                                              vol.get_affine(), [x1, x2, x3])
-
-                # sanity mask: some voxels might have fallen out of business;
-                # and zap'em
-                msk = np.nonzero((y1 >= 0) & (y1 < dim[0]) & (y2 >= 0)
-                                 & (y2 < dim[1]) & (y3 >= 0)
-                                 & (y3 < dim[2]))[0]
-
-                # if mask is too small, then we're screwed anyway
-                if len(msk) < 32:
-                    raise RuntimeError(
-                        ("Almost all voxels have fallen out of the FOV. Only "
-                         "%i voxels survived. Registration can't work." % len(
-                             msk)))
-
-                # warp: resample volume t on this new grid
-                F = ndimage.map_coordinates(V, [y1[msk], y2[msk], y3[msk]],
-                                            order=self.interp, mode='wrap')
-
-                # formulate and solve LS problem for updating p
-                A = A0[msk, ...].copy()
-                b1 = b[msk].copy()
-                sc = np.sum(b1) / np.sum(F)
-                b1 -= F * sc
-                q_update = scipy.linalg.lstsq(np.dot(A.T, A),
-                                              np.dot(A.T, b1))[0]
-
-                # update q
-                q[self.lkp] += q_update
-                rp[t, self.lkp] -= q_update
-
-                # update affine matrix for volume t by applying M_q
-                vol = apply_realignment_to_vol(vol, q)
-
-                # compute convergence criterion variables
-                pss = ss
-                ss = np.sum(b1 ** 2) / len(b1)
-
-                # compute relative gain over last iteration
-                relative_gain = np.abs((pss - ss) / pss
-                                       ) if np.isfinite(pss) else INFINITY
-
-                # verbose
-                token = "\t" + "".join(['%-12.4g ' % z for z in q[self.lkp]])
-                token += '|  %.5g' % relative_gain
-                self._log(token)
-
-                # check whether we've stopped converging altogether
-                if relative_gain < self.tol and countdown == -1:
-                    countdown = 2
-
-                # countdown
-                if countdown != -1:
-                    # converge
-                    if countdown == 0:
-                        break
-                    countdown -= 1
-
-            # what happened ?
-            comments = " after %i iterations" % (iteration + 1)
-            if iteration + 1 == self.n_iterations:
-                comments = "did not coverge" + comments
-            else:
-                if relative_gain < self.tol:
-                    comments = "converged" + comments
-                else:
-                    comments = "stopped converging" + comments
 
-            self._log("")
+        if n_jobs > 1:
```

Comment:
Tip: A special treatment for the case `n_jobs` can be avoided (this will also avoid code duplication). Joblib automatically defaults to sequential execution if `n_jobs = 1`